File: Dijkstra.pl
use strict;
use warnings;

use constant INF => 1_000_000_000;

sub dijkstra {
    my ($graph, $start) = @_;
    my $n = scalar @$graph;

    my @dist = (INF) x $n;
    my @visited = (0) x $n;

    $dist[$start] = 0;

    for (1 .. $n - 1) {
        my $min = INF;
        my $u = -1;

        # Trouver le sommet non visité avec la distance minimale
        for my $i (0 .. $n - 1) {
            if (!$visited[$i] && $dist[$i] < $min) {
                $min = $dist[$i];
                $u = $i;
            }
        }

        last if $u == -1;

        $visited[$u] = 1;

        # Mise à jour des distances
        for my $v (0 .. $n - 1) {
            if (
                !$visited[$v] &&
                $graph->[$u][$v] != 0 &&
                $dist[$u] + $graph->[$u][$v] < $dist[$v]
            ) {
                $dist[$v] = $dist[$u] + $graph->[$u][$v];
            }
        }
    }

    return @dist;
}

# Exemple d'utilisation
my @graph = (
    [0, 3, 0, 0, 5],
    [3, 0, 2, 0, 0],
    [0, 2, 0, 4, 0],
    [0, 0, 4, 0, 1],
    [5, 0, 0, 1, 0],
);

my $start = 0;
my @distances = dijkstra(\@graph, $start);

print "Distances depuis le sommet $start\n";
for my $i (0 .. $#distances) {
    print "Vers $i : $distances[$i]\n";
}

File: dijkstra.rb
INF = 1_000_000_000

def dijkstra(graph, start)
  n = graph.length
  dist = Array.new(n, INF)
  visited = Array.new(n, false)

  dist[start] = 0

  (n - 1).times do
    u = -1
    min = INF

    # Trouver le sommet non visité avec la distance minimale
    (0...n).each do |i|
      if !visited[i] && dist[i] < min
        min = dist[i]
        u = i
      end
    end

    break if u == -1
    visited[u] = true

    # Mise à jour des distances
    (0...n).each do |v|
      if !visited[v] && graph[u][v] != 0 &&
         dist[u] + graph[u][v] < dist[v]
        dist[v] = dist[u] + graph[u][v]
      end
    end
  end

  dist
end

# Exemple d'utilisation
graph = [
  [0, 7, 0, 0, 2],
  [7, 0, 3, 0, 0],
  [0, 3, 0, 4, 0],
  [0, 0, 4, 0, 1],
  [2, 0, 0, 1, 0]
]

start = 0
distances = dijkstra(graph, start)

puts "Distances depuis le sommet #{start}"
distances.each_with_index do |d, i|
  puts "Vers #{i} : #{d}"
end

